server/src/com/thoughtworks/go/config/update/RoleConfigCommand.java

```diff
@@ -63,71 +57,124 @@ public boolean canContinue(CruiseConfig cruiseConfig) {
         return isAuthorized();
     }
 
-    protected boolean isValidForCreateOrUpdate(CruiseConfig preprocessedConfig) {
-        preprocessedRole = findExistingRole(preprocessedConfig);
-        preprocessedRole.validate(null);
-        validate(preprocessedConfig);
+    @Override
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        preprocessedRole = preprocessedConfig.server().security().getRoles().findByName(role.getName());
 
-        if (preprocessedRole.errors().isEmpty()) {
-            preprocessedConfig.server().security().getRoles().validate(null);
+        if (!preprocessedRole.validateTree(validationContextWithSecurityConfig(preprocessedConfig))) {
             BasicCruiseConfig.copyErrors(preprocessedRole, role);
-            return preprocessedRole.getAllErrors().isEmpty() && role.errors().isEmpty();
+            return false;
         }
 
-        BasicCruiseConfig.copyErrors(preprocessedRole, role);
-        return false;
+        return true;
     }
 
-    private void validate(CruiseConfig preprocessedConfig) {
-        if (role instanceof PluginRoleConfig) {
-            PluginRoleConfig role = (PluginRoleConfig) this.role;
-            PluginRoleConfig preprocessedRole = (PluginRoleConfig) this.preprocessedRole;
-
-            SecurityAuthConfig securityAuthConfig = preprocessedConfig.server().security().securityAuthConfigs().find(role.getAuthConfigId());
-            if (securityAuthConfig == null) {
-                role.addError("authConfigId", "No such security auth configuration present " + role.getAuthConfigId());
-                return;
-            }
-            try {
-                ValidationResult result = extension.validateRoleConfiguration(securityAuthConfig.getPluginId(), role.getConfigurationAsMap(true));
-                if (!result.isSuccessful()) {
-                    for (ValidationError validationError : result.getErrors()) {
-                        ConfigurationProperty property = preprocessedRole.getProperty(validationError.getKey());
-                        if (property == null) {
-                            role.addNewConfiguration(validationError.getKey(), false);
-                            preprocessedRole.addNewConfiguration(validationError.getKey(), false);
-                            property = preprocessedRole.getProperty(validationError.getKey());
-                        }
-                        property.addError(validationError.getKey(), validationError.getMessage());
-
-                    }
-                }
-            } catch (PluginNotFoundException e) {
-                role.addError("authConfigId", "Could not find a security authorization config with id '" + role.getAuthConfigId() + "'.");
+    protected ValidationContext validationContextWithSecurityConfig(final CruiseConfig preprocessedConfig) {
```

Comment:
Instead of this mostly empty impementation, should we have a default no-op impl, or just make `ValidationContext` a concrete no-op class instead?